.travis.yml

```diff
@@ -1,8 +1,10 @@
 # Stuff shared by all builds
 install:
-  - python setup.py install
+  - pip install ".[dev]"
 script:
-  - python -m pytest
+  - pytest
+  - flake8 autofocus tests
+  - black autofocus tests --check
```

Comment:
now that this is getting a little more complicated, you might consider a pattern I've really started to like lately and that I have seen other open source projects use. 

In [doppel-cli](https://github.com/jameslamb/doppel-cli/blob/master/.travis.yml#L17), for example, I use `.travis.yml` mainly for _orchestration_ (enumeration and ordering of steps) and I put all of the logic of those steps into shell scripts in a folder like `.ci/`.

This is powerful for a lot of reasons, but I think the two biggest benefits are:

* readability of your `.travis.yml` as your CI gets more complex
* makes it easy for people to run the tests locally (so they can satisfy those items on the PR checklist)

I think you should consider that!

For all of these CI systems, they just run whatever code you give them and fail builds if anything returns a non-zero exit code. So, concretely, I would change the script section to

```
script:
    - ./.ci/test.sh
```

And then make the test script

```
#!/bin/bash

# guarantee script exits with non-zero code
# the first time any of these commands fail
set -e

pytest

flake8 autofocus tests

black autofocus tests \
    --check
```

Reply:
also be sure to `chmod +x ./ci/test.sh` so it's executable with `./`. If you've never seen it before, files in most types of filesystems keep around a bit that tracks whether or not they're executable, so `git` will actually be aware that the file is executable (meaning you won't have to re-run `chmod +x ...` on Travis and people won't have to on their laptops)